### PR TITLE
[evm] fix GetHash function

### PR DESCRIPTION
**Pull request by dustinxie**

### CODE REVIEW

File: blockchain/integrity/integrity_test.go [hunk 1]
```diff
@@ -684,48 +684,52 @@ func addTestingGetBlockHash(t *testing.T, hawaiiHeight uint64, bc blockchain.Blo
...
-	require.NoError(err)
-	acHash4, err := addOneBlock(contract, 4, zero, gasLimit, gasPrice, getBlockHash(1)) // equal to block 2
-	require.NoError(err)
-	acHash5, err := addOneBlock(contract, 5, zero, gasLimit, gasPrice, getBlockHash(3)) // equal to block 1
-	require.NoError(err)
-	acHash6, err := addOneBlock(contract, 6, zero, gasLimit, gasPrice, getBlockHash(2)) // equal to block 3
-	require.NoError(err)
-	acHash7, err := addOneBlock(contract, 7, zero, gasLimit, gasPrice, getBlockHash(6)) // equal to genesis block
-	require.NoError(err)
-
```
Comment (dustinxie) [inline blockchain/integrity/integrity_test.go:703]: move into for loop to make the code more clean

**Comment by Liuhaai** [comment]
Could we test `opBlockhash` in solidity?

**Comment by dustinxie** [comment]
> Could we test `opBlockhash` in solidity?

yes, the test deploys a contract which end up calling `opBlockhash` in solidity: https://github.com/iotexproject/iotex-core/pull/2998/files#diff-376031ecf84904d1ad817660ba4a12fcf9143faed75224a178fdad3ebcefd63eL621-L634

**Review by CoderZhi — COMMENTED** [review]
https://github.com/ethereum/go-ethereum/blob/b1e72f7ea998ad662166bcf23705ca59cf81e925/core/vm/instructions.go#L442-L454

**Comment by dustinxie** [comment]
> https://github.com/ethereum/go-ethereum/blob/b1e72f7ea998ad662166bcf23705ca59cf81e925/core/vm/instructions.go#L442-L454

yes, that's EVM internal implementation, it checks if the requested number is within last 256 blocks, if yes, call `GetHash()`, otherwise set to all-0
so what is the specific thing to do/change?

**Review by CoderZhi — APPROVED** [review]

**Comment by Liuhaai** [comment]
Could you check why is time fixed in L647
`blockTime := time.Unix(1546329600, 0)`

File: blockchain/integrity/integrity_test.go [hunk 2]
```diff
@@ -684,48 +684,52 @@ func addTestingGetBlockHash(t *testing.T, hawaiiHeight uint64, bc blockchain.Blo
...
+		getHashHeight uint64
 	}{
-		{acHash2, 2, 0},
-		{acHash3, 3, 2},
-		{acHash4, 4, 2},
-		{acHash5, 5, 1},
-		{acHash6, 6, 3},
-		{acHash7, 7, 0},
+		{2, 0},
+		{3, 5},
```
Comment (Liuhaai) [inline blockchain/integrity/integrity_test.go:697]: b/c `commitHeight` is added one by one. So there is no need to be set in testdata

File: blockchain/integrity/integrity_test.go [hunk 3]
```diff
@@ -684,48 +684,52 @@ func addTestingGetBlockHash(t *testing.T, hawaiiHeight uint64, bc blockchain.Blo
...
+		{5, 3},
+		{6, 0},
+		{7, 6},
+		{8, 9},
+		{9, 3},
+		{10, 9},
+		{11, 1},
+		{12, 4},
+		{13, 0},
+		{14, 100},
```
Comment (Liuhaai) [inline blockchain/integrity/integrity_test.go:708]: add one more test {15, 15}?

File: action/protocol/execution/evm/evm.go [hunk 4]
```diff
@@ -110,22 +110,31 @@ func newParams(
...
 
 	var getHashFn vm.GetHashFunc
-	if !featureCtx.FixGetHashFnHeight {
+	switch {
+	case featureCtx.CorrectGetHashFn:
 		getHashFn = func(n uint64) common.Hash {
-			hash, err := getBlockHash(stateDB.blockHeight - n)
-			if err != nil {
+			hash, err := getBlockHash(n)
+			if err == nil {
```
[the remark this answers is not in the document]
Comment (dustinxie) [reply action/protocol/execution/evm/evm.go:117]: for clarity would rather keep it same as current code

File: blockchain/integrity/integrity_test.go [hunk 5]
```diff
@@ -684,48 +684,52 @@ func addTestingGetBlockHash(t *testing.T, hawaiiHeight uint64, bc blockchain.Blo
...
 	}
 	for _, test := range tests {
-		r, err := dao.GetReceiptByActionHash(test.acHash, test.commitHeight)
+		h, err := addOneBlock(contract, nonce, zero, gasLimit, gasPrice, getBlockHash(int64(test.getHashHeight)))
```
[the remark this answers is not in the document]
Comment (dustinxie) [reply blockchain/integrity/integrity_test.go:711]: yes, it is the data of tx, but what the tx does is to `getBlockHash`, so i think it's fine?

File: blockchain/integrity/integrity_test.go (hunk 3 above)
Comment (dustinxie) [reply blockchain/integrity/integrity_test.go:708]: added in 2nd commit

File: action/protocol/execution/evm/evm.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (Liuhaai) [reply action/protocol/execution/evm/evm.go:117]: ok. Could you file a new issue to correct the style here?

File: blockchain/integrity/integrity_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (Liuhaai) [reply blockchain/integrity/integrity_test.go:712]: No. the name `getBlockHash` is expected to return a hash. But it returns []byte.
I'd prefer `calldata`

File: action/protocol/execution/evm/evm.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (dustinxie) [reply action/protocol/execution/evm/evm.go:117]: i don't think it's necessary? we have lots of `if err == nil` in the codebase

File: blockchain/integrity/integrity_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (dustinxie) [reply blockchain/integrity/integrity_test.go]: added in 3rd commit

**Review by Liuhaai — APPROVED** [review]

**Pull request merged**